beartype/_util/utilcallable.py

```diff
@@ -19,39 +19,68 @@
-#     Parameters
-#     ----------
-#     func : Callable
-#         Callable to be inspected.
-#
-#     Returns
-#     ----------
-#     str
-#         Either:
-#
-#         * If this callable is physically declared by an uncompiled Python
-#           script or module, the absolute filename of this script or module.
-#         * Else, the placeholder string ``"<string>"`` implying this callable to
-#           have been dynamically declared in-memory.
-#
-#     See Also
-#     ----------
-#     :func:`inspect.getsourcefile`
-#         Inefficient stdlib function strongly inspiring this implementation,
-#         which has been highly optimized for use by the performance-sensitive
-#         :func:`beartype.beartype` decorator.
-#     '''
-#
-#     pass
+def get_callable_filename_or_placeholder(func: Callable) -> str:
+    '''
+    Absolute filename of the uncompiled Python script or module physically
+    declaring the passed callable if any *or* the placeholder string
+    ``"<string>"`` implying this callable to have been dynamically declared
+    in-memory otherwise.
+
+    Parameters
+    ----------
+    func : Callable
+        Callable to be inspected.
+
+    Returns
+    ----------
+    str
+        Either:
+
+        * If this callable is physically declared by an uncompiled Python
+          script or module, the absolute filename of this script or module.
+        * Else, the placeholder string ``"<string>"`` implying this callable to
+          have been dynamically declared in-memory.
+
+    See Also
+    ----------
+    :func:`inspect.getsourcefile`
+        Inefficient stdlib function strongly inspiring this implementation,
+        which has been highly optimized for use by the performance-sensitive
+        :func:`beartype.beartype` decorator.
+    '''
+    if callable(func) is False:
+        raise _BeartypeUtilCallableException(
+            f"Attempted to pass non-callable:{repr(func)}")
+
+   # Grab location of the module declaring class
+    elif isinstance(func, type):
+
+        # Check the Class has a module thats registered with the system.
+        if (getattr(func,"__module__", None) and
+            sys.modules.get(func.__module__) and
+            getattr(sys.modules.get(func.__module__),"__file__", None)):
+            filename = os.path.abspath(sys.modules.get(func.__module__).__file__)
+        # Class was dynamically declared.
+        else:
+            filename = "<string>"
+    else:
+        # Built-in types are callables with no .py source file.
+        if isinstance(func,CallableCTypes):
+            filename = "<builtin>"
+        # Check the function/method contains a code object with the declaring filename.
+        elif (getattr(func, "__code__", None) and
+              getattr(func.__code__, "co_filename", None)):
+            filename = os.path.abspath(func.__code__.co_filename)
+        # Callable was dynamically declared.
+        else:
+            filename = "<string>"
+
+    return filename
```

Comment:
**Superb first draft.** Because the whole point of this exercise that has depleted your essential vitality is absurd efficiency, I have comments that will make you begin to dread life itself:

* **Object attribute lookup is surprisingly expensive,** because it implicitly entails a dictionary lookup and thus call to the `dict.__getitem__()` method. This means *every* `.` operator (e.g., `sys.modules`, `func.__module__`). In Python microoptimization, we like to optimize away repeated object attribute lookup by only performing the lookup once. And by we, I mean me. I am horrible! *Fear my disgusting ear whiskers.*
* The prior point means that it's *much* more efficient to import attributes directly (e.g., `from sys import modules`) rather than access attributes through modules imported directly (e.g., `sys.modules`).
* I'm *pretty* sure (but *not* entirely sure) that all pure-Python callables have code objects. Equivalently, if a function or method isn't C-based, it really had better have a `__code__` attribute. Ergo, `getattr(func, "__code__", None)` *should* be safely reducible to just `func.__code__`.
* **Bound methods wrap unbound functions.** Fortunately, distinguishing between bound methods and unbound functions is trivial and efficient. If it's an instance of `beartype.cave.MethodBoundInstanceOrClassType`, it's a bound method; if it's an instance of `beartype.cave.FunctionType`, it's an unbound function. You are probably thinking: "Will he stop talking?" No! No, I won't. I like talking. This distinction matters, because **only unbound functions have `__code__` attributes.** So how do you get the `__code__` attribute for a bound method? Simple: you get the unbound function wrapped by that bound method first via the `__func__` attribute (e.g., `func.__func__.__code__` for bound methods). *Python is Hell.*
* For efficiency, simplicity, and safety, it's often best to initialize the local variable you *eventually* want to return to a safe fallback first. In this case, that means initializing `filename = '<string>'` **as the first statement of this function.** If you do that, you'll hopefully find that the resulting implementation becomes much simpler, more readable, and possibly even faster. These are all good things! Of course, I might also be wrong and doing that will just destroy the structure of the entire fragile edifice. @leycec bestows "wisdom" even as he rubs his fearsome bushy eyebrows.
* Whenever you see repeated code fragments (e.g., `sys.modules.get(func.__module__)`), you just know you've violated [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself). That's bad, because it's slower than it otherwise could be. Thankfully, we've now initialized `filename = '<string>'` beforehand, which means that whole conditional on line 67 can be unravelled a bit.
* [The `__file__` attribute of *all* modules excluding one edge case that we don't care about is guaranteed to be absolute rather than relative under Python >= 3.4](https://stackoverflow.com/a/22866630/2809027). That means we can drop the first `os.path.abspath()` call, which would just reduce to a noop in 99.99999% of all calls anyway. *Down with the noop!*
* [The same applies to the `co_filename` attribute of code objects under Python >= 3.9](https://bugs.python.org/issue20443). I'm inclined to ignore relative `co_filename` attributes under older Python versions, as (again) this is only needed for non-mission-critical error messages. *Down with Python < 3.9!*
* We can be even cleverer with our `getattr()` calls and pass `filename` rather than `None` as their third parameters, **which makes the code stupidly unreadable but deliciously fast.** I choose deliciously fast. Let someone else maintain this.

All in all, I'm just another thick brick in the wall:

```
    filename = '<string>'

    # Grab location of the module declaring class
    if isinstance(func, type):
        # Check the Class has a module thats registered with the system.
        func_module_name = getattr(func, '__module__', None)

        if func_module_name is not None:
            func_module = modules.get(func_module_name)
            filename = getattr(func_module, '__file__', filename)
    else:
        # Built-in types are callables with no .py source file.
        if isinstance(func, CallableCTypes):
            filename = '<builtin>'
        # Check the function/method contains a code object with the declaring filename.
        else:
            if isinstance(func, MethodBoundInstanceOrClassType):
                func = func.__func__

            filename = getattr(func.__code__, "co_filename", filename)
            
    return filename
```

**Bwahah!** It's as beautiful as a plastic bag fluttering in the wind that reminds me of the ephemerality of my own decaying meat suit. I am crying. :crying_cat_face:

Reply:
>  I'm _pretty_ sure (but _not_ entirely sure) that all pure-Python callables have code objects. Equivalently, if a function or method isn't C-based, it really had better have a `__code__` attribute. Ergo, `getattr(func, "__code__", None)` _should_ be safely reducible to just `func.__code__`.
> 
>  Likewise, _all_ code objects should have `co_filename` attributes. For in-memory callables, the value of that attribute should _already_ be `<string>`. Ergo, `getattr(func.__code__, "co_filename", None)` should also be safely reducible to just `func.__code__.co_filename`.
> 
>  Likewise, _all_ classes should have `__module__` attributes. For in-memory classes, the value of that attribute should _already_ be `None`. Ergo, `getattr(func, "__module__", None)` should also be safely reducible to just `func.__module__`.

>  [The `__file__` attribute of _all_ modules excluding one edge case that we don't care about is guaranteed to be absolute rather than relative under Python >= 3.4](https://stackoverflow.com/a/22866630/2809027). That means we can drop the first `os.path.abspath()` call, which would just reduce to a noop in 99.99999% of all calls anyway. _Down with the noop!_
> 
>  [The same applies to the `co_filename` attribute of code objects under Python >= 3.9](https://bugs.python.org/issue20443). I'm inclined to ignore relative `co_filename` attributes under older Python versions, as (again) this is only needed for non-mission-critical error messages. _Down with Python < 3.9!_

I assumed such, but wasn't sure about bad actors. I'm ever increasing my <s>power 😈 </s> knowledge of OOP and am not aware if someone could, through omission or some Pythonic anti-magic, effectively delete a dunder. (i.e. The dunder difference between Mapping v. MutableMapping) 


>  Sadly, the prior point suggests that tests aren't _quite_ there yet. Tests shouldn't have passed, but they did. I get tired just thinking about tests, so I'll leave this in your ingenious hands. I apologize for everything.

Yeah, I'm still getting the hang of the testing bit. In this particular instance writing the test cases before developing the solution proved, and is proving, to be [challenging to understand](https://youtu.be/3VA9vuId3ds?t=16)

Clean is good! My initial implementation felt a little chunky. I did make one change based on my own interest in Byte-Code/Assembly. It's an odd fascination whose practicality in Python is questionable at best but here we are nonetheless 
At a byte-code level in ***Python 3.9*** this call is roughly 2 instructions longer  than it's equivalent. Why? - the  `is not false` convention uses the `IS_OP` code which requires loading in the value of  `None` to check whether the current value in fact the opposite.
>```
>         if func_module_name is not None:
>             filename = getattr(modules[func_module_name], '__file__', filename)
>```
```
 79     >>   88 LOAD_GLOBAL              3 (isinstance)
             90 LOAD_FAST                0 (func)
             92 LOAD_GLOBAL              9 (MethodBoundInstanceOrClassType)
             94 CALL_FUNCTION            2
             96 LOAD_CONST               6 (None)
             98 IS_OP                    1
            100 POP_JUMP_IF_FALSE      108

```
Contrast that with the piece of code which will throw `False` in the case the value is `None`
```
if func_module_name:
    filename = ....
```

```
 79     >>   88 LOAD_GLOBAL              3 (isinstance)
             90 LOAD_FAST                0 (func)
             92 LOAD_GLOBAL              9 (MethodBoundInstanceOrClassType)
             94 CALL_FUNCTION            2
             96 POP_JUMP_IF_FALSE      104
```
Note: I am not sure what this looks like in a pre-3.9 environment

Reply:
> At a byte-code level in Python 3.9 this call is roughly 2 instructions longer than it's equivalent.

**Oh, shit.** We got a bytecode badass here. It's awesome and horrifying that you went deeper than even I – the self-proclaimed neckbeard on the frozen lake with a bald head, two cats, and no hockey sticks – have gone on microoptimization. And I go deep.

I'm depressed that `func_module_name is not None` is slow. It makes sense, but that doesn't make me any less depressed. This is why we defer to the bytecode badass. I will have further need of your special "services" when we go deep into the bytecode hole with `beartypecache`, a prospective project that will involve cray-cray Abstract Syntax Tree (AST) and bytecode manipulation destined to destroy this timeline. See `codemain.py` if sanity-eroding immolation-by-code-complexity is your thing.